### ISSUE TITLE
Simplify test setup

### ISSUE THREAD

**Issue by xlc** [paritytech/substrate#3335]
They are many duplicated setup code required for any new modules, mostly config `system::Trait`. There must be a way to avoid duplicate this code all over the places.
It also make any modification to `system::Trait` tedious as it have to change every single module test setup.

I am not sure what is the best way, but at least a macro can avoid most of the duplications.

https://github.com/paritytech/substrate/blob/9b3e9f32cd5243d2eeff3d5d8337033aa64c091b/srml/aura/src/mock.rs#L35-L63

**Comment by kianenigma** [comment]
I cannot agree more. The `system::Trait` is becoming a bottleneck. 

Ideally, I imagine this as the least that we can do: 
- Each module must expose its default `Trait` implementation and types. 
- When module x with runtime `Test` wants to have tests, it can implement `Trait` for `Test` or fallback to the default. 

With a bit more effort we can also probably generalize the `Builder()` pattern with proc-macros to enable all tests to have it. 

At some point if I have the time, I would really like to work on this.

**Comment by kianenigma** [comment]
I have a vague guess that maybe @KiChjang would be interested in this.

**Comment by KiChjang** [comment]
Indeed, it looks to me that a derive macro like `DefaultTestSetup` could generate all the necessary code for the default test framework setup. Let me see what I can do here.

**Comment by kianenigma** [comment]
Re-posting my suggestion https://github.com/paritytech/substrate/issues/12130#issuecomment-1282123683 since I think it might be a good starting point for this.

**Comment by kianenigma** [comment]
> The vision in https://github.com/paritytech/substrate/issues/12130#issuecomment-1282123683 can also be achieved with something like this: 

* `#[pallet::genesis_config]` is structurally very similar to a pallet's hypothetical `ExtBuilder`. 
* Why not have `#[pallet::genesis_config]` generate a basic `ExtBuilder` for you, that you can extend?
* crucially, this `ExtBuilder` would have a given `fn build_and_execute(..)` that can automatically call into `try_state` and such for you.
